### PR TITLE
Store clientapplication "last seen" info per privacyIDEA node

### LINKED ISSUE #1819: Store event counters and client application timestamps per node to improve performance of replication
In replicated setups like Galera, we sometimes have problems with multiple nodes trying to write to the same database row at the same time. This happens especially in case of the ``clientapplication`` table, which records the last seen date of the client application and is modified in each ``/validate/check`` request. If counting event handlers are defined, the same happens for the ``eventcounter`` tables. In case of Galera, multiple nodes trying to write to the same row might cause deadlocks (see #1268).

We could try to mitigate the issue by having each privacyIDEA node only write counters to its own private table row. As an example, with ``eventcounter``: We could add a column ``node`` to the table, which holds the name of the privacyIDEA node that the counter belongs to. Assume that we have three privacyIDEA nodes (A, B and C), each connected to its own Galera node.

Then, A updates an event counter like this: ``UPDATE eventcounter SET counter_value = counter_value + 1 WHERE counter_name = 'mycounter' AND node = 'A'``

And B updates an event counter like this: ``UPDATE eventcounter SET counter_value = counter_value + 1 WHERE counter_name = 'mycounter' AND node = 'B'``

C updates an event counter like this: ``UPDATE eventcounter SET counter_value = counter_value + 1 WHERE counter_name = 'mycounter' AND node = 'C'``

Then, when retrieving the event counter value with a given name, we would need to sum up all rows of all nodes: ``SELECT SUM(counter_value) FROM eventcounter WHERE counter_name = 'mycounter'``.

This way, each Galera node would only try to write to its own private row of ``eventcounter``, which would reduce the number of deadlocks.

However, this would only work if each privacyIDEA node has its own Galera node. It wouldn't work, for example, if the privacyIDEA nodes connect to a load balancer that selects the Galera node to use.

### PR DESCRIPTION
**Pull request by fredreichbier**
This also adds a corresponding migration script and tests and reduces the number of deadlocks to zero in the setup mentioned in #1819.

The migration script has two disadvantages:
* It doesn't work on SQLite, because it doesn't allow removing constraints
* Downgrading probably doesn't work, because removing the ``node`` field may violate UNIQUE constraints. But I'm not sure if downgrading currently works at all? :)

Fixes #1819

### CODE REVIEW

File: tests/test_lib_clientapplication.py [hunk 1]
```diff
@@ -38,3 +42,86 @@ def test_01_save_and_get(self):
...
+        self.assertEqual(row3.lastseen, t2)
+        # ... but on pinode2, it was t1!
+        row4 = ClientApplication.query.filter_by(ip="1.2.3.4", clienttype="RADIUS", node="pinode2").one()
+        self.assertEqual(row4.lastseen, t1)
+
+        # check that the apps are returned correctly
+        apps = get_clientapplication(ip="1.2.3.4")
+        self.assertEqual(set(apps.keys()), {"PAM", "RADIUS"})
+        self.assertEqual(apps["PAM"], [{"ip": "1.2.3.4", "hostname": None, "lastseen": t2}])
+        self.assertEqual(apps["RADIUS"], [{"ip": "1.2.3.4", "hostname": None, "lastseen": t2}])
```
Comment (plettich) [inline tests/test_lib_clientapplication.py]: Can we assume, that the dictionary order does not change here?

File: tests/test_lib_clientapplication.py [hunk 2]
```diff
@@ -38,3 +42,86 @@ def test_01_save_and_get(self):
...
+
+        apps = get_clientapplication(group_by="ip")
+        self.assertEqual(set(apps.keys()), {"1.2.3.4", "2.3.4.5"})
+        self.assertEqual(len(apps["1.2.3.4"]), 2)
+        self.assertIn({"clienttype": "PAM", "hostname": None, "lastseen": t2}, apps["1.2.3.4"])
+        self.assertIn({"clienttype": "RADIUS", "hostname": None, "lastseen": t2}, apps["1.2.3.4"])
+        self.assertEqual(apps["2.3.4.5"], [{"clienttype": "PAM", "hostname": None, "lastseen": t1}])
```
Comment (plettich) [inline tests/test_lib_clientapplication.py]: same here. Or does `assertEqual()` ignore dictionary order?

File: tests/test_lib_clientapplication.py (hunk 1 above)
Comment (fredreichbier) [reply tests/test_lib_clientapplication.py]: I think even though dictionaries officially remember the insertion order in Python 3.7, this doesn't affect equality checks: so ``{"a": "b", "c":"d"}`` will always be equal to ``{"c": "d", "a": "b"}``, and this is what ``assertEqual`` checks :)
